### PR TITLE
feat(button-group): web component created

### PR DESCRIPTION
**Pull request by RaphaelAmadeu**
### Related Ticket(s)

#3492 

### Description

Web-component created, the same width functionality will be added in another issue.

### CODE REVIEW

File: packages/web-components/src/components/button-group/__tests__/button-group.test.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      },
+    },
+  });
+
+describe('dds-button-group', function() {
+  it('renders properly', async function() {
+    render(template(), document.body);
+    await Promise.resolve(); // Update cycle for `<dds-button-group>`
+    expect(document.body.querySelector('dds-button-group')).toMatchSnapshot({ mode: 'shadow' });
+    expect(document.body.querySelector('dds-button-group-item')).toMatchSnapshot({ mode: 'shadow' });
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/__tests__/button-group.test.ts:40]: Two snapshots cannot be tested in one test case, so could you please divide this?

File: packages/web-components/src/components/button-group/button-group.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    `;
+  }
+
+  /**
+   * Handler for @slotchange, set the first button-group-item to kind tertiary and primary for the remaining ones
+   *
+   * @private
+   */
+  private _handleSlotChange(event: Event) {
+    this.childItems = (event.target as HTMLSlotElement)
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group.ts:62]: Does any code refer to `childItems`? If not, probably a local variable?

File: packages/web-components/src/components/button-group/button-group.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      .assignedNodes({ flatten: true })
+      .filter(elem => (elem as HTMLElement).localName === (this.constructor as typeof DDSButtonGroup).buttonGroupItemSelector);
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group.ts:64]: We have an API to test if an element matches a selector:

```suggestion
      .filter(elem => (elem as HTMLElement).matches((this.constructor as typeof DDSButtonGroup).buttonGroupItemSelector));
```

File: packages/web-components/src/components/button-group/button-group.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  static get stableSelector() {
+    return `${ddsPrefix}--button-group`;
+  }
+
+  static get buttonGroupItemSelector() {
+    return `${ddsPrefix}-button-group-item`;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  protected render() {
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group.ts:48]: To align to the rest of the codebase:

```suggestion
  render() {
```

File: packages/web-components/src/components/button-group/button-group.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  /**
+   * Handler for @slotchange, set the first button-group-item to kind tertiary and primary for the remaining ones
+   *
+   * @private
+   */
+  private _handleSlotChange(event: Event) {
+    this.childItems = (event.target as HTMLSlotElement)
+      .assignedNodes({ flatten: true })
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group.ts:63]: We are not deep-walking `<slot>`s, right...?

```suggestion
      .assignedNodes()
```

File: packages/web-components/src/components/button-group/button-group-item.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  static get stableSelector() {
+    return `${ddsPrefix}--button-group-item`;
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    if (!this.hasAttribute('role')) {
+      this.setAttribute('role', 'listitem');
+    }
+  }
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group-item.ts:34]: Style nit, the rest of the codebase put `static` fields/methods the bottom:

```suggestion
  connectedCallback() {
    super.connectedCallback();
    if (!this.hasAttribute('role')) {
      this.setAttribute('role', 'listitem');
    }
  }

  static get stableSelector() {
    return `${ddsPrefix}--button-group-item`;
  }

  static styles = styles; // `styles` here is a `CSSResult` generated by custom WebPack loader
```

**Review by asudoh — COMMENTED** [review]
Great progress @RaphaelAmadeu!

File: packages/web-components/src/components/button-group/button-group.ts (hunk 3 above)
Comment (RaphaelAmadeu) [reply packages/web-components/src/components/button-group/button-group.ts]: @asudoh I've implemented this, but I had to add a verification for the existence of `matches` function in the element, as the text nodes returned by `slot` don't have this function built into them.

File: packages/web-components/src/components/button-group/button-group.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  static get buttonGroupItemSelector() {
+    return `${ddsPrefix}-button-group-item`;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  render() {
+    return html`
+      <ol>
+        <slot @slotchange="${this._handleSlotChange}" />
+      </ol>
```
Comment (asudoh) [inline packages/web-components/src/components/button-group/button-group.ts:48]: My apologies for slipping this through the crack in my earlier review. To [avoid duplicate element between the host and shadow DOM](https://github.com/carbon-design-system/ibm-dotcom-library/blob/v1.11.0-rc.0/packages/web-components/docs/coding-conventions.md#custom-element-itself-as-an-eleement). Please make sure adding `list` role in a similar manner as [`<dds-button-group-item>`](https://github.com/carbon-design-system/ibm-dotcom-library/blob/b502a3239c3497b6863bc8d99064890e3030208c/packages/web-components/src/components/button-group/button-group-item.ts#L29-L31). Also please apply the same style as `.bx--buttongroup` to the host. Thanks!

```suggestion
      <slot @slotchange="${this._handleSlotChange}" />
```

**Review by asudoh — COMMENTED** [review]
Almost there @RaphaelAmadeu!

File: packages/web-components/src/components/button-group/button-group.ts (hunk 7 above)
Comment (RaphaelAmadeu) [reply packages/web-components/src/components/button-group/button-group.ts]: @asudoh Just updated it, thanks for your reviews!

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @RaphaelAmadeu!

File: packages/web-components/src/components/button-group/__stories__/README.stories.mdx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+### HTML
+
+```html
+<dds-button-group>
+  <dds-button-group-item>Lorem ipsum</dds-button-group-item>
+  <dds-button-group-item>Dolor sit amet</dds-button-group-item>
+</dds-button-group>
+```
```
Comment (jeffchew) [inline packages/web-components/src/components/button-group/__stories__/README.stories.mdx:40]: @RaphaelAmadeu looks good! Can you update the README to show an example of including attributes like `href`, how to add an icon, etc?

Comment (RaphaelAmadeu) [reply packages/web-components/src/components/button-group/__stories__/README.stories.mdx:40]: @jeffchew I've updated it on my latest pr

**Review by jeffchew — APPROVED** [review]
LGTM!

**Pull request merged**